src/compiler/checker.ts

```diff
@@ -35560,9 +35560,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             else if (type && !hasExplicitReturn) {
                 // minimal check: function has syntactic return type annotation and no explicit return statements in the body
                 // this function does not conform to the specification.
-                error(errorNode, Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);
+                if (strictNullChecks) {
+                    error(errorNode, Diagnostics.A_function_whose_declared_type_is_neither_undefined_void_nor_any_must_return_a_value);
+                }
+                else {
+                    error(errorNode, Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);
+                }
             }
-            else if (type && strictNullChecks && !isTypeAssignableTo(undefinedType, type)) {
+            else if (type && strictNullChecks) {

```

Comment:
@jakebailey @DanielRosenwasser The removal of this assignability check has caused a regression. This example now errors when it shouldn't:

```ts
function fx2(x: boolean): unknown {
  if (x) {
    return "hello";
  }
};
```

Reply:
I noticed the regression here: https://github.com/microsoft/TypeScript/issues/53515#issuecomment-1484215602

Reply:
Funny. We were just talking about adding this check back again in https://github.com/microsoft/TypeScript/pull/53490

Reply:
I'll look into why it's not working as we expected.

Reply:
Oh, no, that PR was exactly to fix the issue you mentioned. So it should be fixed in main now.

Reply:
Checked, and it is fixed, yep.